sv.c

```diff
@@ -4337,17 +4332,21 @@ Perl_sv_setsv_flags(pTHX_ SV *dsv, SV* ssv, const I32 flags)
         }
         if (!SvROK(ssv))
             goto undef_sstr;
+#if NVSIZE > IVSIZE
         if (dtype < SVt_PV && dtype != SVt_IV)
             sv_upgrade(dsv, SVt_IV);
```

Comment:
Yeah, this existing code is a bit curiously written, presumable for historical reasons. (Maybe it predates bodyless IVs? And RVs no longer having their own type?)

Prior to this PR, regardless of NV size, any NULL will be upgraded to an IV via bodyless tricks, and this will happen in the special "bodyless" casing at the top of the function.

With bodyless NV:
* Prior to this PR, the special "bodyless" casing didn't handle a source IV and a destination NV (and vice versa), so this code was reachable.
* With this PR, this code is no longer reachable - and such builds will fail the `#if` test.

With arena body NVs, this statement is still reachable, because the special bodyless casing will only handle NULL/IVs in such builds. However, the code could be more clearly rewritten now as:
 
```
#if NVSIZE > IVSIZE
        if (dtype == SVt_NV)
            sv_upgrade(dsv, SVt_PVIV);
```
And we only get here if `SvROK(ssv)`. I'm unclear on whether a NV is meant to be capable of storing a RV or not. If not (which seems likely), I don't know if the correct thing to do is excise this code or swap the `sv_upgrade` for a `croak`. It's currently still here out of caution for this one potential use-case, basically.

Reply:
I dont remember from 5 years ago enough, but I think P5P opinion is to never implement SV body downgrades.  SVRV is long gone as an independent type code and its just a SVIV with a flag. Perhaps 5.6/5.8 era perl bless related struct fields got moved from SVRV out wrapper to SVMG real obj, and SVRV lost its purpose. Someone could argue Perl's ```SvTYPE()``` is a solid U8 of metadata, and needs to represent "defined contents" of the SV more than ```SvTYPE()```'s current malloc book keeping purpose.  Perl SV head U32 or U24 bitfield, encode so much metadata, and the behaviors modeled are not` ``adjacent``` (fuzzy word) in the bitfield,  so U32->U8 split bitwise tricks are hard. So P5 core dev cant do ``` for(i=6;i;i--) {sv_flags = sv_flags >>1; if(){}}``` type things. other than IPNR/IPNRp-Ok() in byte 2 which is linear.

A certain perl person, not me, not naming them, did argue with me that P5P keeps their metadata too many pointers/layers deep. And totally wasted the "gift" of the SV head's 4th slot. This person wants to see var len malloced C structs at the start of HvARRAY and AvARRAY and PV buffer/VLAs, and not the current and day 1 perl 5 setup of array ptrs being pure array ptrs.

PV HEK and UTF8PV sneak in metadata into the "array", but these 3 APIs keep it on the end at random offsets. its not a struct.

He also proposed deleting field 4 in the SV head, and ```->sv_any``` becomes malloced, and toss the legacy P5 arena API except for SV heads.

I wouldnt vote yes if I had to make an official vote with his design proposals above, but this person is correct IMO that the HV * layout and HE*  has way too many ptr derefs. Ive seen at YAPCs, commercial grade code, that does PP Arrays with PP const subs labels for array indexes. Its not a blog post semi-joke anymore.

Blessed arrays with integer labeled methods, yes there is commercial Perl code with that design (C++ objects in PP) running, but its a minority, profilers would show I/O, SQL DBs, OSes, and ping time matter more.

LVALUE tie/LVALUE magic vs SV .h macros vs perlguts POD is most difficult part of the SV scalar half of the type system to learn or debug. LVALUE set magic firing on lvalue assignment from an XSUB. Plus LVALUE type is rarely seen I think, most apps and most scalar SVs on planet earth only peak at SVt_PVMG, and SVt_PVLV is IV NV PV buffered string type, but it sits out of place in the type system.

SVGV is below SVPVLV. So if there will be bugs related to P5's exactly architecture PP types in implemented in C types, they will happen in SVPVLV type, CPAN, p5p, or private XS code. Useless SvUPGRADE() calls to type PVLV is thought experiment and how many asserts, SEGVs, and real bugs come up.

Reply:
I too am interested in the potential for alternatives to the current type bodies (or even doing "fat heads" for e.g. AV and HV"), but not as part of this PR.